File: web/php/class/Router.php
<?php

class Router
{
  private $routes;

  public function __construct($routes)
  {
    $this->routes = $routes;
  }

  public function detect($path) {
    $context = null;

    foreach ($this->routes as $route) {
      $ptn = "@${route['match']}@";

      if (preg_match($ptn, $path, $matches)) {
        foreach ( $matches as $key => $val ) {
          if (is_string($key)) {
            $route[$key] = $val;
          }
        }

        $context = $route;

        if (array_key_exists('uri', $context)) {
          $context['uri'] = $path;
        }

        break;
      }
    }

    return $context;
  }
}

File: ruby/router.rb
class Router
  def initialize(routes)
    @routes = routes.map do |route|
      route[:_names] = route['match'].scan(/\?\<([^>]+)\>/).flatten
      route
    end
  end

  def detect(path)
    match = nil
    route = @routes.find do |route|
      pattern = Regexp.new(route['match'])
      match = pattern.match(path)
    end

    return nil unless match

    res = {}

    route[:_names].each do |name|
      name = name.to_sym

      begin
        match[name]
      rescue
      else
        res[name.to_s] = match[name]
      end
    end

    route = route.dup
    route.delete(:_names)

    res['uri'] = path unless res['uri']
    res['status'] = 200 unless res['status']

    res.merge(route)
  end
end